### ISSUE TITLE
Fix bug in `staticHTML` whereby sample names in documents are sometimes wrong

### ISSUE THREAD

**Issue by mikej888** [riboviz/riboviz#372]
See https://github.com/riboviz/riboviz/issues/193#issuecomment-784624146 and https://github.com/riboviz/riboviz/issues/355#issuecomment-863519851

**Comment by mikej888** [comment]
For the `staticHTML` error I edited `vignette_config.yaml` to have 4 samples:
```
  WTnone: SRR1042855_s1mi.fastq.gz
  WT3AT: SRR1042864_s1mi.fastq.gz
  NotHere: example_missing_file.fastq.gz # Test case for missing file
=>
  WT1: SRR1042855_s1mi.fastq.gz
  WT2: SRR1042864_s1mi.fastq.gz
  WT3: SRR1042855_s1mi.fastq.gz
  WT4: SRR1042864_s1mi.fastq.gz
```
I ran the workflow but it failed...
```console
$ nextflow run -ansi-log false prep_riboviz.nf -params-file vignette/vignette_config.yaml 
N E X T F L O W  ~  version 20.04.1
Launching `prep_riboviz.nf` [determined_yonath] - revision: f12ad0b907
samples_dir: .
organisms_dir: .
data_dir: .
[a3/d2702c] Submitted process > buildIndicesrRNA (yeast_rRNA)
[35/18a84e] Submitted process > cutAdapters (WT3)
[02/bc240a] Submitted process > buildIndicesORF (YAL_CDS_w_250)
[86/e8dcbb] Submitted process > cutAdapters (WT1)
[3e/96fd7c] Submitted process > cutAdapters (WT4)
[84/ab9772] Submitted process > cutAdapters (WT2)
[15/40958b] Submitted process > createVizParamsConfigFile
[5b/1542bf] Submitted process > hisat2rRNA (WT3)
[2d/c21c2e] Submitted process > hisat2rRNA (WT1)
[a9/d69f5c] Submitted process > hisat2rRNA (WT2)
[48/632c9b] Submitted process > hisat2rRNA (WT4)
[47/6d55e8] Submitted process > hisat2ORF (WT1)
[5c/e80ac6] Submitted process > hisat2ORF (WT3)
[81/2b33a1] Submitted process > trim5pMismatches (WT1)
[f2/c84fb4] Submitted process > trim5pMismatches (WT3)
[8c/3d5fe5] Submitted process > samViewSort (WT1)
[49/efde2f] Submitted process > samViewSort (WT3)
[22/c1ce66] Submitted process > outputBams (WT3)
[68/87455e] Submitted process > outputBams (WT1)
[4b/5d7543] Submitted process > makeBedgraphs (WT3)
[46/212c79] Submitted process > bamToH5 (WT3)
[db/b1b2f8] Submitted process > bamToH5 (WT1)
[d6/ddad35] Submitted process > makeBedgraphs (WT1)
[34/6f4651] Submitted process > hisat2ORF (WT2)
[ca/ade434] Submitted process > hisat2ORF (WT4)
[ea/5bedfc] Submitted process > trim5pMismatches (WT2)
[96/b500a7] Submitted process > trim5pMismatches (WT4)
[e6/0491ac] Submitted process > samViewSort (WT2)
[51/d0f3d8] Submitted process > samViewSort (WT4)
[48/701d7d] Submitted process > outputBams (WT4)
[38/318c2a] Submitted process > outputBams (WT2)
[de/6747be] Submitted process > makeBedgraphs (WT2)
[89/80600c] Submitted process > bamToH5 (WT4)
[29/197d9e] Submitted process > makeBedgraphs (WT4)
[4e/97dd36] Submitted process > bamToH5 (WT2)
[42/a88195] Submitted process > generateStatsFigs (WT1)
[b3/ceafa0] Submitted process > generateStatsFigs (WT3)
[8d/a9cead] Submitted process > generateStatsFigs (WT4)
[d8/18e68e] Submitted process > generateStatsFigs (WT2)
Finished processing sample: WT3
[d6/e26386] Submitted process > renameTpms (WT3)
[bd/6c08e6] Submitted process > staticHTML (WT3)
Finished processing sample: WT1
[2c/6f899c] Submitted process > renameTpms (WT1)
[5f/f3ab67] Submitted process > staticHTML (WT1)
Finished processing sample: WT4
[a0/302859] Submitted process > renameTpms (WT4)
[86/5273d9] Submitted process > staticHTML (WT4)
Finished processing sample: WT2
[00/c0b8dd] Submitted process > renameTpms (WT2)
[52/b5201e] Submitted process > staticHTML (WT2)
[65/76fe79] Submitted process > collateTpms (WT3, WT1, WT4, WT2)
Finished visualising sample: WT3
Finished visualising sample: WT1
[3f/764845] Submitted process > countReads
Finished visualising sample: WT4
Error executing process > 'staticHTML (WT2)'

Caused by:
  Process `staticHTML (WT2)` terminated with an error exit status (1)
...
```
Samples WT1, WT3 and WT4 succeeded but sample WT2 fails during `staticHTML` even though it is the same sample as sample WT4! So I compared the commands run for WT4 and WT2:

```
$ cat work/86/5273d9bbdf9177ef7028201d891aa6/.command.sh
#!/bin/bash -ue
Rscript -e "rmarkdown::render('/home/ubuntu/riboviz/rmarkdown/AnalysisOutputs.Rmd',params = list(verbose='FALSE', yamlfile='$PWD/config.yaml', sampleid='WT4', three_nucleotide_periodicity_data_file = '$PWD/3nt_periodicity.tsv', gene_position_length_counts_5start_file = '$PWD/gene_position_length_counts_5start.tsv', read_length_data_file='$PWD/read_lengths.tsv', pos_sp_rpf_norm_reads_data_file='$PWD/pos_sp_rpf_norm_reads.tsv' , gene_read_frames_filtered_data_file='$PWD/3ntframe_bygene_filtered.tsv', codon_ribodens_gathered_file='$PWD/codon_ribodens_gathered.tsv', sequence_features_file='$PWD/sequence_features.tsv' ), output_format = 'html_document', output_file = '$PWD/WT4_output_report.html')"

$ cat work/52/b5201e9d0b81d96cc7f1d4fdd9a85a/.command.sh 
#!/bin/bash -ue
Rscript -e "rmarkdown::render('/home/ubuntu/riboviz/rmarkdown/AnalysisOutputs.Rmd',params = list(verbose='FALSE', yamlfile='$PWD/config.yaml', sampleid='WT2', three_nucleotide_periodicity_data_file = '$PWD/3nt_periodicity.tsv', gene_position_length_counts_5start_file = '$PWD/gene_position_length_counts_5start.tsv', read_length_data_file='$PWD/read_lengths.tsv', pos_sp_rpf_norm_reads_data_file='$PWD/pos_sp_rpf_norm_reads.tsv' , gene_read_frames_filtered_data_file='$PWD/3ntframe_bygene_filtered.tsv', codon_ribodens_gathered_file='$PWD/codon_ribodens_gathered.tsv', sequence_features_file='$PWD/sequence_features.tsv' ), output_format = 'html_document', output_file = '$PWD/WT2_output_report.html')"
```
and they are the same, except for the sample ID, as expected.

So I compared the logs...

```console
$ cat work/86/5273d9bbdf9177ef7028201d891aa6/.command.log 
...
output file: AnalysisOutputs.knit.md

/usr/bin/pandoc +RTS -K512m -RTS AnalysisOutputs.utf8.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /home/ubuntu/riboviz/work/86/5273d9bbdf9177ef7028201d891aa6/WT4_output_report.html --self-contained --variable bs3=TRUE --standalone --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /home/ubuntu/R/x86_64-pc-linux-gnu-library/3.6/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --variable theme=bootstrap --include-in-header /tmp/RtmpY7pI6Q/rmarkdown-str1c4941f82a22.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' 

Output created: WT4_output_report.html
Warning messages:
1: Removed 4 rows containing non-finite values (stat_smooth). 
2: Removed 4 rows containing missing values (geom_point). 
3: Removed 24 rows containing non-finite values (stat_smooth). 
4: Removed 24 rows containing missing values (geom_point). 
```
```console
$ cat work/52/b5201e9d0b81d96cc7f1d4fdd9a85a/.command.log 
...
output file: AnalysisOutputs.knit.md

/usr/bin/pandoc +RTS -K512m -RTS AnalysisOutputs.utf8.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /home/ubuntu/riboviz/work/52/b5201e9d0b81d96cc7f1d4fdd9a85a/WT2_output_report.html --self-contained --variable bs3=TRUE --standalone --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /home/ubuntu/R/x86_64-pc-linux-gnu-library/3.6/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --variable theme=bootstrap --include-in-header /tmp/RtmpbcCrSy/rmarkdown-str1c8c8402c4d.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' 
pandoc: AnalysisOutputs.utf8.md: openFile: does not exist (No such file or directory)
Error: pandoc document conversion failed with error 1
In addition: Warning messages:
1: Removed 4 rows containing non-finite values (stat_smooth). 
2: Removed 4 rows containing missing values (geom_point). 
3: Removed 24 rows containing non-finite values (stat_smooth). 
4: Removed 24 rows containing missing values (geom_point). 
Execution halted
```
The failure for WT2 relates to
```
pandoc: AnalysisOutputs.utf8.md: openFile: does not exist (No such file or directory)
```
So the next step is to find out why that file isn't there. `AnalysisOutputs.utf8.md` seems to be auto-generated somewhere as an intermediate file. It is also unclear _where_ this file is created. It doesn't include a sample-specific ID. It _might_ be the case that it is created not in the task-specific `work` subfolder, but in `rscripts` or a folder common to all samples which _might_ explain why this error arises (`staticHTML` for WT2 comes to use that file just as `staticHTML` for another sample deletes it). It also _might_ explain why samples end up in unexpected HTML reports.

Next stop, let's look at what `rmarkdown::render` does!

**Comment by mikej888** [comment]
From https://pkgs.rstudio.com/rmarkdown/reference/render.html, `rmarkdown::render` supports two parameters that we don't currently use but which are of interest:

> intermediates_dir	
>
> Intermediate files directory. If a path is specified then intermediate files will be written to that path. If NULL, intermediate files are written to the same directory as the input file.
>
> knit_root_dir	
>
> The working directory in which to knit the document; uses knitr's root.dir knit option. If NULL then the behavior will follow the knitr default, which is to use the parent directory of the document.

Given that `AnalysisOutputs.md` is in `rscripts/` the comments on these parameters imply that any intermediate files will be created in `rscripts/` also. 

A web search provided a useful command, `watch -n 0.1 "ls -ltr"`, which displays the results of `ls -ltr` every 0.1 seconds, so I monitored `rmarkdown/` folder using:

```console
$ watch -n 0.1 "ls -ltr" rmarkdown/
```
I reran the workflow. Two files prefixed by AnalysisOutputs repeatedly appeared and disappeared when rerunning the workflow. Also the workflow completed successfully but with WT1's HTML document having WT3 in it's 'Sample Name'

So I'd more strongly suspect now that the reason for the mixed up sample IDs, and, for me, the failed run, is due to `rmarkdown::render` creating intermediate files with the sample name for different samples in the same directory and these are colliding.

Next step, try out those additional `rmarkdown::render` parameters.

**Comment by FlicAnderson** [comment]
@mikej888 you're a wonder to behold! The intermediate files sounds like a really strong candidate for the culprit, based on your detective work!

**Comment by mikej888** [comment]
From my last run of the workflow:
```
[53/fe92d4] Submitted process > staticHTML (WT1)
[be/d09105] Submitted process > staticHTML (WT2)
[0e/383ea2] Submitted process > staticHTML (WT3)
[2b/365d98] Submitted process > staticHTML (WT4)
```
Change into `staticHTML (WT1)` directory:
```console
$ cd work/53/fe92d4a65a7145248505b0fd39a9c5/
$ watch -n 0.1 "ls -ltr"
```
In another window, run `.command.sh`:
```console
$ cd work/53/fe92d4a65a7145248505b0fd39a9c5/
$ bash .command.sh
```
Files flicker up in the `watch` window:
```console
drwxr-xr-x 3 ubuntu ubuntu    4096 Jun 18 05:17 WT1_output_report_files
```
Edit `.command.sh` and add the following to `rmarkdown::render`'s arguments:
```
intermediates_dir = '$PWD'
```
then rerun:
```console
$ bash .command.sh
```
Again, files flicker up in the `watch` window:
```console
-rw-r--r-- 1 ubuntu ubuntu    6897 Jun 18 05:17 AnalysisOutputs.Rmd
drwxr-xr-x 3 ubuntu ubuntu    4096 Jun 18 05:17 WT1_output_report_files
...this time with others too fast to copy...
```
Edit `prep_riboviz.nf` and add:
```
      script += "intermediates_dir = '\$PWD', "
```
to `staticHTML`'s `shell` commands to build the `rmarkdown::render` invocation.

Rerun workflow while `watch`-ing `rmarkdown/`

No additional files appear.

Check HTMLs shows the correct sample IDs in each one

Rerun the workflow again shows correct sample IDs in the correct HTMLs.

**Comment by mikej888** [comment]
Added 1 line fix to `prep_riboviz.nf` in new branch [fix-static-html-372](https://github.com/riboviz/riboviz/tree/fix-static-html-372) (branched from `develop`).

Tested on Ubuntu 18, Ubuntu 20 and CentOS 7 VMs - reran default integration tests and also ran on`WT1-4` sample configuration as above to check for wrong sample IDs, there were none :-)

@flicanderson please give it a go yourself, thanks.

**Comment by mikej888** [comment]
Summary...

When `rmarkdown::render`, invoked by `staticHTML`, renders a document it creates intermediate files in the same directory as the document being rendered and with the same prefix as the document. For us these were always the directory `rmarkdown` and the intermediate files  `AnalysisOutputs.<SUFFIX>`.

Depending on when sample-specific invocations of `staticHTML` ran this could result in the same intermediate file being used for more than one sample. This why there where occasionally HTML documents with the wrong sample data in them, why sometimes `staticHTML` for a sample would fail (the `rmarkdown::render` invocation for one sample would delete an intermediate file when another `rmarkdown::render` was underway)  also why it was difficult to reproduce this error every time!

Using an `intermediates_dir` parameter in `rmarkdown::render` allows the directory for these intermediate files to be specified. This is set to be the sample-specific Nextflow task subdirectories (under `work/`) so that the intermediate files for each sample-specific invocation of `rmarkdown::render`, by `staticHTML`, don't clash with each other.

**Comment by FlicAnderson** [comment]
> Added 1 line fix to `prep_riboviz.nf` in new branch [fix-static-html-372](https://github.com/riboviz/riboviz/tree/fix-static-html-372) (branched from `develop`).
> 
> Tested on Ubuntu 18, Ubuntu 20 and CentOS 7 VMs - reran default integration tests and also ran on`WT1-4` sample configuration as above to check for wrong sample IDs, there were none :-)
> 
> @FlicAnderson please give it a go yourself, thanks.

Thanks, I'll give this a whirr after the L-Sc_2014 I'm re-running locally finishes up shortly (it's at countReads and it'd be a shame to cancel it so close to the end).

**Comment by FlicAnderson** [comment]
@mikej888 I ran G-Hv_2020 (3 samples) through your `fix-static-html-372` branch and the html outputs look good - names correct, locations correct, plots as they should be.  :muscle: 

I think (unless you want me to test further on other datasets) that this is good to be merged to `develop` (and `main`).

**Comment by mikej888** [comment]
Merged into develop (373) and on into main (#374).

**Comment by kavousan** [comment]
Brilliant job Mike, well done.
